system/Encryption/Encryption.php

```diff
@@ -0,0 +1,902 @@
+	/**
+	 * List of available modes
+	 *
+	 * @var	array
+	 */
+	protected $_modes = array(
+		'mcrypt' => array(
+			'cbc' => 'cbc',
+			'ecb' => 'ecb',
+			'ofb' => 'nofb',
+			'ofb8' => 'ofb',
+			'cfb' => 'ncfb',
+			'cfb8' => 'cfb',
+			'ctr' => 'ctr',
+			'stream' => 'stream'
+		),
+		'openssl' => array(
+			'cbc' => 'cbc',
+			'ecb' => 'ecb',
+			'ofb' => 'ofb',
+			'cfb' => 'cfb',
+			'cfb8' => 'cfb8',
+			'ctr' => 'ctr',
+			'stream' => '',
+			'xts' => 'xts'
+		)
+	);
+
+	/**
+	 * List of supported HMAC algorithms
+	 *
+	 * name => digest size pairs
+	 *
+	 * @var	array
+	 */
+	protected $_digests = array(
+		'sha224' => 28,
+		'sha256' => 32,
+		'sha384' => 48,
+		'sha512' => 64
+	);
+
+	/**
+	 * mbstring.func_overload flag
+	 *
+	 * @var	bool
+	 */
+	protected static $func_overload;
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Class constructor
+	 *
+	 * @param	array	$params	Configuration parameters
+	 * @return	void
+	 * 
+	 * @throws \CodeIgniter\Encryption\EncryptionException
+	 */
+	public function __construct($config = null)
+	{
+		if (empty($config))
+			$config = new \Config\Encryption();
+		$this->config = $config;
+
+		$params = (array) $this->config;
+
+		$this->_drivers = array(
+			'mcrypt' => defined('MCRYPT_DEV_URANDOM'),
+			'openssl' => extension_loaded('openssl')
+		);
+
+		if (!$this->_drivers['mcrypt'] && !$this->_drivers['openssl'])
+		{
+			throw new EncryptionException('Unable to find an available encryption driver.');
+		}
+
+		isset(self::$func_overload) OR self::$func_overload = (extension_loaded('mbstring') && ini_get('mbstring.func_overload'));
+		$this->initialize($params);
+
+		if (!isset($this->_key) && self::strlen($key = $this->config->key) > 0)
+		{
+			$this->_key = $key;
+		}
+
+		log_message('info', 'Encryption Class Initialized');
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Initialize
+	 *
+	 * @param	array	$params	Configuration parameters
+	 * @return	CI_Encryption
+	 * 
+	 * @throws \CodeIgniter\Encryption\EncryptionException
+	 */
+	public function initialize(array $params)
+	{
+		if (!empty($params['driver']))
+		{
+			if (isset($this->_drivers[$params['driver']]))
+			{
+				if ($this->_drivers[$params['driver']])
+				{
+					$this->_driver = $params['driver'];
+				} else
+				{
+					throw new EncryptionException("Driver '" . $params['driver'] . "' is not available.");
+				}
+			} else
+			{
+				throw new EncryptionException("Unknown driver '" . $params['driver'] . "' cannot be configured.");
+			}
+		}
+
+		if (empty($this->_driver))
+		{
+			$this->_driver = ($this->_drivers['openssl'] === TRUE) ? 'openssl' : 'mcrypt';
+
+			log_message('debug', "Encryption: Auto-configured driver '" . $this->_driver . "'.");
+		}
+
+		empty($params['cipher']) && $params['cipher'] = $this->_cipher;
+		empty($params['key']) OR $this->_key = $params['key'];
+		$this->{'_' . $this->_driver . '_initialize'}($params);
```

Comment:
Remove the _ from method names. It's a leftover relic from PHP4 days that isn't needed now that we have actual protected methods. The only time I've been using that now is for abstract methods that a child class would implement.